deepinv/sampling/sde.py

```diff
@@ -0,0 +1,408 @@
+
+        .. note::
+           The :attr:`rng` attribute must be initialized for this method to work properly.
+        """
+        self.rng_manual_seed(seed)
+        return torch.empty_like(input).normal_(generator=self.rng)
+
+
+class DiffusionSDE(nn.Module):
+    r"""
+    Forward-time and Reverse-time Diffusion Stochastic Differential Equation with parameterized drift term.
+
+    The forward SDE is defined by:
+
+    .. math::
+        d\, x_{t} = f(x_t, t) d\,t + g(t) d\, w_{t}.
+
+    This forward SDE can be reversed by the following SDE running backward in time:
+
+    .. math::
+        d\, x_{t} = \left( f(x_t, t) - g(t)^2 \nabla \log p_t(x_t) \right) d\,t + g(t) d\, w_{t}.
+
+    There also exists a deterministic probability flow ODE whose trajectories share the same marginal distribution as the SDEs:
+
+    .. math::
+        d\, x_{t} = \left( f(x_t, t) - \frac{1}{2} g(t)^2 \nabla \log p_t(x_t) \right) d\,t.
+
+
+    The score function can be computed using Tweedie's formula, given a MMSE denoiser :math:`D`:
+
+    .. math::
+        \nabla \log p_{t}(x) = \left( D(x, \sigma(t)) - x \right) / \sigma(t)^2
+
+    where :math:`sigma(t)` is the noise level at time :math:`t`, which can be accessed through the attribute :meth:`sigma_t`
+
+    Default parameters correspond to the `Ornstein-Uhlenbeck process <https://en.wikipedia.org/wiki/Ornstein%E2%80%93Uhlenbeck_process>`_ SDE, defined in the time interval `[0,1]`.
+
+    :param callable drift: a time-dependent drift function :math:`f(x, t)` of the forward-time SDE.
+    :param callable diffusion: a time-dependent diffusion function :math:`g(t)` of the forward-time SDE.
+    :param callable denoiser: a pre-trained MMSE denoiser which will be used to approximate the score function by Tweedie's formula.
+    :param bool rescale: a boolean indicating whether to rescale the input and output of the denoiser to match the scale of the drift.
+                        Should be set to `True` if the denoiser was trained on `[0,1]`. Default to `False`.
+
+    :param bool use_backward_ode: a boolean indicating whether to use the deterministic probability flow ODE for the backward process.
+    :param torch.Generator rng: pseudo-random number generator for reproducibility.
+    :param torch.dtype dtype: data type of the computation, except for the `denoiser` which will be always compute in `float32`.
+        We recommend using `torch.float64` for better stability and less numerical error when solving the SDE in discrete time, since
+        most computation cost is from evaluating the `denoiser`, which will be always computed in `float32`.
+    :param torch.device device: device on which the computation is performed.
+    """
+
+    def __init__(
+        self,
+        drift: Callable = lambda x, t: -x,  # Default to Ornstein-Uhlenbeck process
+        diffusion: Callable = lambda t: math.sqrt(2.0),
+        use_backward_ode=False,
+        denoiser: nn.Module = None,
```

Comment:
when we extend the class to inverse problems, here we should add a NoisyDataFidelity class input

Reply:
Yes you're reason. I think that we have 2 options:
- add the `Score` input argument instead, which will be the unconditional score `\nabla \log p_t(x_t)` or the conditional score `\nabla \log p_t(x_t \vert y)`, but then we should define these scores properly outside the `DiffusionSDE`, which can be annoying when switching between the time and the noise level (which is dependent to the underlying SDE) (which is done by the function `sigma_t`).
- add an extra input argument `NoisyDataFidelity` when creating the SDE for the conditional sampling, but it is not used here. But I think that we have no problem when inheriting the ConditionalDiffusionSDE from the DiffusionSDE with an extra input argument. 
I think I prefer the option 2.

Reply:
yes, I think it is better to be more specific and ask directly for a `NoisyDataFidelity`